From 9bed27eac196fd8eae5d12fd90f1f098d8945f6a Mon Sep 17 00:00:00 2001
From: Kiet Pham <kiet.pham.xb@renesas.com>
Date: Sun, 4 May 2025 14:13:36 +0700
Subject: [PATCH] video: Add GRAY10_LE64 support

This add a 10bit variant of gray scale packed into 64bits little endian
words. The MSB 4 bits are padding and should be ignored.

This format is used by RZ Renesas SoC only and it has not been identified with
the FOURCC.

Ref: https://github.com/GStreamer/gst-plugins-base/commit/f7a27d

Signed-off-by: Kiet Pham <kiet.pham.xb@renesas.com>
---
 gst-libs/gst/video/video-format.c | 80 +++++++++++++++++++++++++++++++
 gst-libs/gst/video/video-format.h | 11 ++++-
 gst-libs/gst/video/video-info.c   |  5 ++
 3 files changed, 95 insertions(+), 1 deletion(-)

diff --git a/gst-libs/gst/video/video-format.c b/gst-libs/gst/video/video-format.c
index ac37a75..8dec5ea 100644
--- a/gst-libs/gst/video/video-format.c
+++ b/gst-libs/gst/video/video-format.c
@@ -4787,6 +4787,84 @@ pack_GRAY10_LE32 (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
   }
 }

+#define PACK_GRAY10_LE64 GST_VIDEO_FORMAT_AYUV64, unpack_GRAY10_LE64, 1, pack_GRAY10_LE64
+static void
+unpack_GRAY10_LE64 (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
+    gpointer dest, const gpointer data[GST_VIDEO_MAX_PLANES],
+    const gint stride[GST_VIDEO_MAX_PLANES], gint x, gint y, gint width)
+{
+  gint i;
+  const guint64 *restrict sy = GET_PLANE_LINE (0, y);
+  guint16 *restrict d = dest;
+  gint num_words = (width + 5) / 6;
+
+  /* Y data is packed into little endian 64bit words, with the 4 MSB being
+   * padding.
+   * -> padding | Y1 | Y2 | Y3 | Y4 | Y5 | Y6
+   */
+
+  for (i = 0; i < num_words; i++) {
+    gint num_comps = MIN (6, width - i * 6);
+    guint pix = i * 6;
+    gsize doff = pix * 4;
+    gint c;
+    guint64 Y;
+
+    Y = GST_READ_UINT64_LE (sy + i);
+
+    for (c = 0; c < num_comps; c++) {
+      guint16 Yn;
+
+      /* For Y, we simply read 10 bit and shift it out */
+      Yn = (Y & 0x03ff) << 6;
+      Y >>= 10;
+
+      if (G_UNLIKELY (pix + c < x))
+        continue;
+
+      if (!(flags & GST_VIDEO_PACK_FLAG_TRUNCATE_RANGE))
+        Yn |= Yn >> 10;
+
+      d[doff + 0] = 0xffff;
+      d[doff + 1] = Yn;
+      d[doff + 2] = 0x8000;
+      d[doff + 3] = 0x8000;
+
+      doff += 4;
+    }
+  }
+}
+
+static void
+pack_GRAY10_LE64 (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
+    const gpointer src, gint sstride, gpointer data[GST_VIDEO_MAX_PLANES],
+    const gint stride[GST_VIDEO_MAX_PLANES], GstVideoChromaSite chroma_site,
+    gint y, gint width)
+{
+  gint i;
+  guint64 *restrict dy = GET_PLANE_LINE (0, y);
+  const guint16 *restrict s = src;
+  gint num_words = (width + 5) / 6;
+
+  for (i = 0; i < num_words; i++) {
+    gint num_comps = MIN (6, width - i * 6);
+    guint pix = i * 6;
+    gsize soff = pix * 4;
+    gint c;
+    guint64 Y = 0;
+    guint64 ds = 0;
+
+    for (c = 0; c < num_comps; c++) {
+      ds = s[soff + 1] >> 6;
+      Y |= ds << (10 * c);
+
+      soff += 4;
+    }
+
+    GST_WRITE_UINT64_LE (dy + i, Y);
+  }
+}
+
 #define PACK_NV12_10LE32 GST_VIDEO_FORMAT_AYUV64, unpack_NV12_10LE32, 1, pack_NV12_10LE32
 static void
 unpack_NV12_10LE32 (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
@@ -5670,6 +5748,8 @@ static const VideoFormat formats[] = {
   MAKE_RGBA_LE_PACK_FORMAT (BGR10A2_LE, "raw video", DPTH10_10_10_2, PSTR4444,
       PLANE0,
       OFFS0, SUB4444, PACK_BGR10A2_LE),
+  MAKE_GRAY_C_LE_FORMAT (GRAY10_LE64, "raw video", DPTH10, PSTR0, PLANE0, OFFS0,
+      SUB4, PACK_GRAY10_LE64),
 };

 static GstVideoFormat
diff --git a/gst-libs/gst/video/video-format.h b/gst-libs/gst/video/video-format.h
index 30189b2..080a233 100644
--- a/gst-libs/gst/video/video-format.h
+++ b/gst-libs/gst/video/video-format.h
@@ -118,6 +118,7 @@ G_BEGIN_DECLS
  * @GST_VIDEO_FORMAT_Y410: packed 4:4:4 YUV, 10 bits per channel(A-V-Y-U...) (Since: 1.16)
  * @GST_VIDEO_FORMAT_VUYA: packed 4:4:4 YUV with alpha channel (V0-U0-Y0-A0...) (Since: 1.16)
  * @GST_VIDEO_FORMAT_BGR10A2_LE: packed 4:4:4 RGB with alpha channel(B-G-R-A), 10 bits for R/G/B channel and MSB 2 bits for alpha channel (Since: 1.16)
+ * @GST_VIDEO_FORMAT_GRAY10_LE64: 10-bit grayscale, packed into 64bit words (4 bits padding)
  *
  * Enum value describing the most common video formats.
  *
@@ -211,6 +212,14 @@ typedef enum {
   GST_VIDEO_FORMAT_Y410,
   GST_VIDEO_FORMAT_VUYA,
   GST_VIDEO_FORMAT_BGR10A2_LE,
+  /**
+   * GST_VIDEO_FORMAT_GRAY10_LE64:
+   *
+   * Grayscale 10bit litle endian packed in 64-bit data (4 bits padding).
+   *
+   * Added by Renesas
+   */
+  GST_VIDEO_FORMAT_GRAY10_LE64,
 } GstVideoFormat;

 #define GST_VIDEO_MAX_PLANES 4
@@ -562,7 +571,7 @@ gconstpointer  gst_video_format_get_palette          (GstVideoFormat format, gsi
   "A420_10LE, A422_10BE, A422_10LE, A444_10BE, A444_10LE, NV61, P010_10BE, " \
   "P010_10LE, IYU2, VYUY, GBRA, GBRA_10BE, GBRA_10LE, BGR10A2_LE, GBR_12BE, GBR_12LE, " \
   "GBRA_12BE, GBRA_12LE, I420_12BE, I420_12LE, I422_12BE, I422_12LE, " \
-  "Y444_12BE, Y444_12LE, GRAY10_LE32, NV12_10LE32, NV16_10LE32, NV12_10LE40 }"
+  "Y444_12BE, Y444_12LE, GRAY10_LE32, NV12_10LE32, NV16_10LE32, NV12_10LE40, GRAY10_LE64 }"

 /**
  * GST_VIDEO_CAPS_MAKE:
diff --git a/gst-libs/gst/video/video-info.c b/gst-libs/gst/video/video-info.c
index 7aed055..0fec71b 100644
--- a/gst-libs/gst/video/video-info.c
+++ b/gst-libs/gst/video/video-info.c
@@ -1088,6 +1088,11 @@ fill_planes (GstVideoInfo * info)
       info->offset[0] = 0;
       info->size = info->stride[0] * GST_ROUND_UP_2 (height);
       break;
+    case GST_VIDEO_FORMAT_GRAY10_LE64:
+      info->stride[0] = (width + 5) / 6 * 8;
+      info->offset[0] = 0;
+      info->size = info->stride[0] * GST_ROUND_UP_2 (height);
+      break;
     case GST_VIDEO_FORMAT_NV12_10LE32:
       info->stride[0] = (width + 2) / 3 * 4;
       info->stride[1] = info->stride[0];
--
2.25.1

